File: src/stxm/absorption.py
import numpy as np
import periodictable
from periodictable import xsf

HC_EV_CM = 1.2398e-4


def mass_absorption_cm2_per_g(
    formula_str: str,
    energy_eV: np.ndarray,
    density_g_cm3: float | None = 1,
) -> np.ndarray:
    """
    Mass absorption coefficient (cm^2/g) from periodictable beta (imaginary part of n).

    Parameters
    ----------
    formula_str : str
        Chemical formula, e.g. "C", "H2O", "CaCO3".
    energy_eV : np.ndarray
        Photon energies in eV.
    density_g_cm3 : float, optional
        Density in g/cm^3. If None, uses formula default.

    Returns
    -------
    np.ndarray
        Mass absorption coefficient mu/rho in cm^2/g at each energy.
    """
    compound = (
        periodictable.formula(formula_str, density=density_g_cm3)
        if density_g_cm3 is not None
        else periodictable.formula(formula_str)
    )
    rho = 1
    energy_keV = np.asarray(energy_eV, dtype=float) / 1000.0
    n = xsf.index_of_refraction(compound, energy=energy_keV, density=rho)
    n = np.atleast_1d(n)
    beta = -np.imag(n)
    lam_cm = HC_EV_CM / np.asarray(energy_eV, dtype=float)
    mu_per_cm = 4 * np.pi * beta / np.atleast_1d(lam_cm)
    return mu_per_cm / rho


def fit_bare_atom_background(
    energy_eV: np.ndarray,
    OD: np.ndarray,
    mu_rho_cm2_per_g: np.ndarray,
    n_low: int = 5,
    n_high: int = 5,
) -> tuple[float, float, np.ndarray, np.ndarray]:
    """
    Fit OD = scale * mu_periodictable + const on first n_low and last n_high energies.

    Parameters
    ----------
    energy_eV : np.ndarray
        Energy in eV (same length as OD and mu_rho).
    OD : np.ndarray
        Optical density (ln I0/I).
    mu_rho_cm2_per_g : np.ndarray
        Mass absorption from periodictable at same energies.
    n_low : int
        Number of points at lowest energies to use in fit.
    n_high : int
        Number of points at highest energies to use in fit.

    Returns
    -------
    scale : float
        Multiplicative scale factor.
    const : float
        Constant background.
    OD_bare : np.ndarray
        Fitted bare-atom curve: scale * mu_rho + const.
    mask_fit : np.ndarray
        Boolean mask of points used in the fit.
    """
    n = len(energy_eV)
    n_low = min(n_low, n)
    n_high = min(n_high, n)
    idx_low = np.arange(n_low) if n_low > 0 else np.array([], dtype=int)
    idx_high = np.arange(n - n_high, n) if n_high > 0 else np.array([], dtype=int)
    idx_fit = np.concatenate([idx_low, idx_high])
    mask_fit = np.zeros(n, dtype=bool)
    if idx_fit.size > 0:
        mask_fit[idx_fit] = True

    mu_fit = mu_rho_cm2_per_g[idx_fit]
    od_fit = OD[idx_fit]

    if n_low == 0:
        scale = np.linalg.lstsq(mu_fit[:, np.newaxis], od_fit, rcond=None)[0][0]
        const = 0.0
    else:
        A = np.column_stack([mu_fit, np.ones_like(mu_fit)])
        scale, const = np.linalg.lstsq(A, od_fit, rcond=None)[0]

    OD_bare = scale * mu_rho_cm2_per_g + const
    return scale, const, OD_bare, mask_fit


def od_to_beta(
    energy_eV: np.ndarray,
    OD: np.ndarray,
    thickness_cm: float,
) -> np.ndarray:
    """
    Convert optical density to beta (imaginary part of refractive index n = 1 - delta - i*beta).

    OD = mu * t, mu = 4*pi*beta/lambda, so beta = OD * lambda / (4*pi*t).

    Parameters
    ----------
    energy_eV : np.ndarray
        Photon energy in eV.
    OD : np.ndarray
        Optical density ln(I0/I).
    thickness_cm : float
        Sample thickness in cm.

    Returns
    -------
    np.ndarray
        Beta (dimensionless) at each energy.
    """
    if thickness_cm <= 0:
        raise ValueError("thickness_cm must be positive")
    lam_cm = HC_EV_CM / np.asarray(energy_eV, dtype=float)
    lam_cm = np.atleast_1d(lam_cm)
    od = np.atleast_1d(OD).astype(float)
    beta = od * lam_cm / (4 * np.pi * thickness_cm)
    return beta

File: src/lib/stxm/absorption.ts
/**
 * NEXAFS absorption and bare-atom normalization (Beer-Lambert OD, CXRO-backed mass absorption).
 *
 * Bare-atom tabulated optical constants follow the same convention as the Python widget:
 * `periodictable.xsf.index_of_refraction` (CXRO-based beta), with
 * mu/rho = (4*pi*beta/lambda) for unit mass density. This matches the STXM notebook pipeline in
 * `stxm.absorption` and differs from xray-atlas Henke f2 mixing
 * (mu/rho = 2*r_e*lambda*N_A*f2/M) only in the tabulation source, not in the step-edge fit:
 * OD ~ scale * mu_bare + const on pre/post-edge windows, then norm. mass abs = (OD - const) / scale.
 */

/** Planck constant times c in eV*cm (wavelength in cm = HC_EV_CM / energy_eV). */
export const HC_EV_CM = 1.2398e-4;

export type BareAtomFitResult = {
  scale: number;
  offset: number;
  odBare: number[];
  fitMask: boolean[];
};

/**
 * Mass absorption coefficient mu/rho (cm^2/g) from tabulated beta: mu = 4*pi*beta/lambda.
 */
export function massAbsorptionFromBeta(energyEv: number[], beta: number[]): number[] {
  return energyEv.map((energy, index) => {
    const lamCm = HC_EV_CM / energy;
    const b = beta[index] ?? 0;
    return (4 * Math.PI * b) / lamCm;
  });
}

/**
 * Fit OD = scale * mu_rho + offset using the lowest `nLow` and highest `nHigh` energy points.
 */
export function fitBareAtomBackground(
  energyEv: number[],
  od: number[],
  muRho: number[],
  nLow = 5,
  nHigh = 5,
): BareAtomFitResult {
  const n = energyEv.length;
  if (n === 0 || od.length !== n || muRho.length !== n) {
    throw new Error("energy_eV, OD, and mu_rho must have the same positive length");
  }
  const nLowUse = Math.min(nLow, n);
  const nHighUse = Math.min(nHigh, n);
  const idxLow = Array.from({ length: nLowUse }, (_, index) => index);
  const idxHigh = Array.from({ length: nHighUse }, (_, index) => n - nHighUse + index);
  const idxFit = [...idxLow, ...idxHigh.filter((index) => !idxLow.includes(index))];
  const fitMask = Array.from({ length: n }, () => false);
  for (const index of idxFit) {
    fitMask[index] = true;
  }

  const muFit = idxFit.map((index) => muRho[index]!);
  const odFit = idxFit.map((index) => od[index]!);

  let scale: number;
  let offset: number;
  if (nLowUse === 0) {
    scale = lstsqOneColumn(muFit, odFit);
    offset = 0;
  } else {
    [scale, offset] = lstsqTwoColumn(muFit, odFit);
  }

  const odBare = muRho.map((mu) => scale * mu + offset);
  return { scale, offset, odBare, fitMask };
}

/**
 * CXRO bare-atom normalized mass absorption (g/cm^2) and uncertainties from OD and tabulated mu/rho.
 */
export function normalizedMassAbsorptionAtEnergies(
  energyEv: number[],
  od: number[],
  odErr: number[],
  muRho: number[],
  options?: { nLow?: number; nHigh?: number; fitOffset?: boolean },
): { values: number[]; errors: number[]; scale: number; offset: number } {
  const nLow = options?.fitOffset === false ? 0 : (options?.nLow ?? 5);
  const nHigh = options?.nHigh ?? 5;
  const { scale, offset } = fitBareAtomBackground(energyEv, od, muRho, nLow, nHigh);
  const scaleSafe = scale !== 0 ? scale : 1;
  const values = od.map((value) => (value - offset) / scaleSafe);
  const errors = odErr.map((value) => Math.abs(value / scaleSafe));
  return { values, errors, scale: scaleSafe, offset };
}

/**
 * Convert optical density to beta (imaginary part of n) at thickness `thicknessCm`.
 */
export function odToBeta(energyEv: number[], od: number[], thicknessCm: number): number[] {
  if (thicknessCm <= 0) {
    throw new Error("thicknessCm must be positive");
  }
  return energyEv.map((energy, index) => {
    const lamCm = HC_EV_CM / energy;
    return ((od[index] ?? 0) * lamCm) / (4 * Math.PI * thicknessCm);
  });
}

/**
 * Uncertainty in beta propagated from OD uncertainty at fixed thickness.
 */
export function odErrToBetaErr(
  energyEv: number[],
  odErr: number[],
  thicknessCm: number,
): number[] {
  if (thicknessCm <= 0) {
    throw new Error("thicknessCm must be positive");
  }
  return energyEv.map((energy, index) => {
    const lamCm = HC_EV_CM / energy;
    return ((odErr[index] ?? 0) * lamCm) / (4 * Math.PI * thicknessCm);
  });
}

function lstsqOneColumn(x: number[], y: number[]): number {
  let sumXx = 0;
  let sumXy = 0;
  for (let index = 0; index < x.length; index += 1) {
    const xi = x[index]!;
    sumXx += xi * xi;
    sumXy += xi * (y[index] ?? 0);
  }
  if (sumXx === 0) {
    return 0;
  }
  return sumXy / sumXx;
}

function lstsqTwoColumn(x: number[], y: number[]): [number, number] {
  let s00 = 0;
  let s01 = 0;
  let s11 = 0;
  let t0 = 0;
  let t1 = 0;
  for (let index = 0; index < x.length; index += 1) {
    const xi = x[index]!;
    const yi = y[index] ?? 0;
    s00 += xi * xi;
    s01 += xi;
    s11 += 1;
    t0 += xi * yi;
    t1 += yi;
  }
  const det = s00 * s11 - s01 * s01;
  if (Math.abs(det) < 1e-30) {
    return [lstsqOneColumn(x, y), 0];
  }
  const scale = (s11 * t0 - s01 * t1) / det;
  const offset = (s00 * t1 - s01 * t0) / det;
  return [scale, offset];
}
